File: app/src/main/java/tech/pauly/findapet/discover/AnimalListAdapter.java
package tech.pauly.findapet.discover;

import android.databinding.DataBindingUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import tech.pauly.findapet.R;

public class AnimalListAdapter extends RecyclerView.Adapter<AnimalListItemViewHolder> {

    private List<AnimalListItemViewModel> animalItems = new ArrayList<>();

    @Inject
    AnimalListAdapter() {}

    @Override
    public AnimalListItemViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        return new AnimalListItemViewHolder(DataBindingUtil.inflate(inflater, R.layout.item_animal_list, parent, false));
    }

    @Override
    public void onBindViewHolder(AnimalListItemViewHolder holder, int position) {
        holder.bind(animalItems.get(position));
    }

    @Override
    public int getItemCount() {
        return animalItems.size();
    }

    public void setAnimalItems(List<AnimalListItemViewModel> items) {
        animalItems.addAll(items);
        notifyDataSetChanged();
    }
}

File: app/src/main/java/tech/pauly/findapet/discover/AnimalListAdapter.kt
package tech.pauly.findapet.discover

import android.databinding.DataBindingUtil
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup

import java.util.ArrayList

import javax.inject.Inject

import tech.pauly.findapet.R

open class AnimalListAdapter @Inject
internal constructor() : RecyclerView.Adapter<AnimalListItemViewHolder>() {

    open var animalItems = ArrayList<AnimalListItemViewModel>()
        set(value) {
            field.addAll(value)
            notifyDataSetChanged()
        }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AnimalListItemViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        return AnimalListItemViewHolder(DataBindingUtil.inflate(inflater, R.layout.item_animal_list, parent, false))
    }

    override fun onBindViewHolder(holder: AnimalListItemViewHolder, position: Int) {
        holder.bind(animalItems[position])
    }

    override fun getItemCount(): Int {
        return animalItems.size
    }

    open fun clearAnimalItems() {
        animalItems.clear()
        notifyDataSetChanged()
    }

    open fun addAnimalItem(item: AnimalListItemViewModel) {
        animalItems.add(item)
        notifyDataSetChanged()
    }
}
